py/objstr.c

```diff
@@ -748,23 +768,11 @@ static mp_obj_t str_finder(size_t n_args, const mp_obj_t *args, int direction, b
 
     GET_STR_DATA_LEN(args[0], haystack, haystack_len);
     GET_STR_DATA_LEN(args[1], needle, needle_len);
+    size_t sub_len;
+    const byte *start = get_substring_data(args[0], n_args - 2, args + 2, &sub_len);
 
-    const byte *start = haystack;
-    const byte *end = haystack + haystack_len;
-    if (n_args >= 3 && args[2] != mp_const_none) {
-        start = str_index_to_ptr(self_type, haystack, haystack_len, args[2], true);
-    }
-    if (n_args >= 4 && args[3] != mp_const_none) {
-        end = str_index_to_ptr(self_type, haystack, haystack_len, args[3], true);
-    }
-
-    if (end < start) {
-        goto out_error;
```

Comment:
Are you sure removing this check is correct?  Granted the tests all still pass, but maybe the test suite misses something?

For example the following doesn't currently work correctly under MP:
```
"a".index("", 2)
```
The change here doesn't fix that (and fixing that is probably out of scope of this PR), but that's an example of something that's currently not tested.

In other words, does the change here have equivalent behaviour to before the change?

Reply:
I believe it is made redundant by the MAX(l line in get_substring_data.

Reply:
I think I found a case which works prior to this PR, but fails with it:
```py
"abc".find("", 1, 0)
```
That would return `-1` prior to this PR because `end < start` and it does the `goto out_error`.

But with this PR it truncates the string to search to the empty string and then successfully finds the empty string within that, returning `1`.

Reply:
You're right of course, now that im back at the computer and can check. I also found another example which is broken on this PR (and also master): `"abc".find("", 5, 6)` returns `3` in both versions of MP and `-1` on cpython. The problem in this case is the use of `is_slice = true` in `str_index_to_ptr()`. I'll take a deeper look. Thanks for picking this up. I'll also add all these cases to the tests.

Reply:
I suggest pulling out the optimisation commit to a separate PR.  So this PR is just the first commit, improving startswith/endswith.  That change is a clear improvement and easier to argue it's "bug free".

Reply:
OK. Shall do. Some further investigation shows strange behaviour in cpython for edge cases where the search string is `""`. I'll document them in the new PR, rather than clutter up this PR.

Reply:
Done.